marlon/defender_agents/defender.py

```diff
@@ -0,0 +1,97 @@
+
+        def get_node_from_action(node_from_action: int):
+            """Converts from action number to node ID."""
+            return list(environment.network.nodes)[node_from_action]
+
+        def get_node_info(node_id: model.NodeID):
+            """Gets node info from node ID."""
+            return environment.get_node(node_id)
+
+        def get_firewall_port_name_from_action(port_name_from_action: int):
+            """Gets the name of the firewall port from the constant firewall rule list."""
+            return self.firewall_rule_list[port_name_from_action]
+
+        def get_service_port_name_from_action(node_id: model.NodeID, port_name_from_action: int):
+            """Gets the service port name from the given node."""
+            node_info = get_node_info(node_id)
+            return node_info.services[port_name_from_action]
+             
+        def block_traffic(node_id: model.NodeID, port_name: model.PortName, incoming: bool):
+            """Blocks traffic on a node to or from a port with port_name."""
+            node_data = environment.get_node(node_id)
+            node_info = get_node_info(node_id)
+            rules = node_data.firewall.incoming if incoming else node_data.firewall.outgoing
+            matching_rules = [r for r in rules if r.port == port_name]
+            if matching_rules:
+                for rule in matching_rules:
+                    node_info.firewall.incoming.remove(rule) if incoming else node_info.firewall.outgoing.remove(rule)
+        
+        def allow_traffic(node_id: model.NodeID, port_name: model.PortName, incoming: bool):
+            """Creates a new firewall rule if one does not exist."""
+            node_data = environment.get_node(node_id)
+            node_info = get_node_info(node_id)
+            rules = node_data.firewall.incoming if incoming else node_data.firewall.outgoing
+            matching_rules = [r for r in rules if r.port == port_name]
+            if not matching_rules:
+                rule_to_add = cyberDefender.model.FirewallRule(port = port_name, permission=cyberDefender.model.RulePermission.ALLOW)
+                node_info.firewall.incoming.append(rule_to_add) if incoming else node_info.firewall.incoming.append(rule_to_add)
+        
+        # If the action is invalid, the list will be empty. In this case the defender will skip its turn.
+        if len(self.next_action) == 0 :
+            return
+        # If the action is a reimage, reimage the node.
+        if self.next_action[0] == 0:
+            actions.reimage_node(get_node_from_action(self.next_action[1]))
+        
+        # If the action is a block traffic.
+        elif self.next_action[0] == 1:
+            node_id = get_node_from_action(self.next_action[2])
+            incoming = bool(self.next_action[4])
+            port_name = get_firewall_port_name_from_action(self.next_action[3])
+            block_traffic(node_id, port_name, incoming)
```

Comment:
I was expecting these functions to be called by the DefenderWrapper directly, instead of being encoded, passed here, decoded, and performed. Much like the `is_defender_action_valid` you wrote. It may allow you to remove the `self.cyber_env._CyberBattleEnv__defender_agent.next_action = action` hack. 

It looks pretty good aside from that hack though, so if that doesn't go away feel free to keep this.

Reply:
As we talked about, I will keep this the way it is until we have the observations and rewards set up, then we will split the attacker and defender completely apart (which will use this change).